### PR TITLE
f08: resolve osx link constant issues

### LINKED ISSUE #4374: bug/jenkins: F08 failures  on osx
```
Test Name | Duration | Age
summary_junit_xml. - ./f08/pt2pt/statusesf08 1 | 72 ms | 2
summary_junit_xml. - ./f08/coll/vw_inplacef08 4 | 0.1 sec | 2
summary_junit_xml. - ./f08/coll/red_scat_blockf08 4 | 89 ms | 2
summary_junit_xml. - ./f08/coll/nonblocking_inpf08 4 | 83 ms | 2
summary_junit_xml. - ./f08/datatype/structf 2 | 0.11 sec | 2
summary_junit_xml. - ./f08/rma/aintf08 2 | 73 ms | 2
summary_junit_xml. - ./f08/topo/dgraph_unwgtf90 4 | 90 ms
```

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

The old mechanism uses separate symbols, e.g. `MPIR_C_MPI_UNWEIGHTED`, as a "link associated" variable between Fortran and C. However, this mechanism depends on how the system linker and may break, for example, on current osx. Instead, we can simply implement utility C functions to get the actual constant, e.g. `MPI_UNWEIGHTED` at runtime. It is more robust and removes the extra complexity of maintaining separate exposed variables.

Other constants, such as `MPI_ARGV_NULL` are simple constants. The old mechanism initializes the corresponding `MPIR_C_MPI_ARGV_NULL` at compile time. While works, this is based on the knowledge/assumption that these constants are defined in `mpi.h`. In this PR, we convert those to use the C utility functions as well for simplification and consistency.

Fixes #4374

[skip warnings]
## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
compiler:intel

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by raffenet** [comment]
The Jenkins results don't contain f08 tests.

This will probably need a review from Cray before we can merge, as they were the original contributors of this code, and helped write it up in our paper. We should also test with the Cray `ftn` compiler.

**Comment by raffenet** [comment]
For reference: https://www.mcs.anl.gov/papers/P5139-0514.pdf section 3.3 describes the named constant implementation.

**Comment by hzhou** [comment]
> The Jenkins results don't contain f08 tests.

Only the tests of `ch3-osx` and `ch4-asan/ubsan` enables `f08`. The former uses `gcc-10` and the latter use `gcc-9`. The f08 tests are there.

**Comment by hzhou** [comment]
> This will probably need a review from Cray before we can merge, as they were the original contributors of this code, and helped write it up in our paper. We should also test with the Cray `ftn` compiler.

Sure, who should we ask for review? Let me ask on the slack.

File: src/binding/fortran/use_mpi_f08/mpi_f08_link_constants.f90
Comment (SteveO-Cray) [inline src/binding/fortran/use_mpi_f08/mpi_f08_link_constants.f90:29]: Are MPI_F08_STATUS_IGNORE and MPI_F08_STATUSES_IGNORE broken similar to MPIR_C_MPI_STATUS_IGNORE and MPIR_C_MPI_STATUSES_IGNORE?  I know MPICH does not use them, but what if the user does?

**Review by SteveO-Cray — COMMENTED** [review]
This hopefully will leave a general comment.

I think the Fortran/C interface routines look fine, but I have sent a sample to Bill Long to check the Fortran, I'm not good at Fortran 2008 either.

The changes in general look good to me, but I don't know python or the python binding script and don't have the time right now to figure it out (and I'm out Friday).  I will try and look at this more while we wait for a response from Bill.

My two other questions should not be considered blocking issues.

If you are in a hurry for some reason to merge this PR, I would say merge it (i.e. I approve) if you can build MPICH and pass the F08 regression tests.  It would be preferable if you could pass using at least 2 different compilers that support Fortran 2008.

File: test/mpi/configure.ac
Comment (hzhou) [reply test/mpi/configure.ac:103]: Thanks for the review! It is not dumb at all. How to match the test suite to the mpich build is tricky. The issue we were having is even though mpich is built with f08, but the tests is not enabled unless we hard code the option into Jenkins' job. We have been missing the test coverage because of this.

Later in the configure there is code to check whether the compiler actually has `2008` support:
https://github.com/pmodels/mpich/blob/26cc6a9ea874b9588dc181ac70bfe4161b57493f/test/mpi/configure.ac#L1411-L1413
So if the compiler is too "old" (e.g. gcc-8), the tests will be disabled later.

File: src/binding/fortran/use_mpi_f08/mpi_f08_link_constants.f90
Comment (hzhou) [reply src/binding/fortran/use_mpi_f08/mpi_f08_link_constants.f90:29]: Both `MPI_F08_STATUS_IGNORE` and `MPI_F08_STATUSES_IGNORE ` are constants (`1`) defined in `mpi.h` and they are statically initialized in `src/binding/fortran/use_mpi_f08/cdesc.c`. It worked, but it relied on the value being constants defined in `mpi.h`.

`MPI_F08_STATUS_IGNORE ` and `MPI_F08_STATUSES_IGNORE ` works the same way as `MPIR_C_MPI_UNWEIGHTED` and `MPIR_C_MPI_WEIGHTS_EMPTY`, but since the symbols on C side is missing `MPICH_API_PUBLIC`, thus not exposed, it is likely broken. But there is nothing checks it.

**Comment by hzhou** [comment]
test:mpich/custom/pipeline
label: osx
compiler: intel

**Comment by hzhou** [comment]
test:mpich/custom/pipeline
label: osx
compiler: intel

**Comment by SteveO-Cray** [comment]
FYI - I am still waiting on a confirmation from Bill Long about the F08 code.  I didn't send him enough info first round (my mistake).  I unfortunately did not get back to this PR, will try to get to it yet today.

**Comment by hzhou** [comment]
> FYI - I am still waiting on a confirmation from Bill Long about the F08 code. I didn't send him enough info first round (my mistake). I unfortunately did not get back to this PR, will try to get to it yet today.

No worries. Meanwhile, I noticed that on `osx` we can't even resolve `MPI_IN_PLACE` and `MPI_BOTTOM`. Those two symbols can't be fixed in the same way as the rest of the symbols since they are directly passed to `wrappers_c`. I am working on the last commit to see if that is due to dynamic linker issue. You can ignore the last commit in your review for now.

**Comment by hzhou** [comment]
test:mpich/custom/pipeline
label: osx
compiler: intel

**Comment by hzhou** [comment]
test:mpich/custom/pipeline
label: osx
compiler: intel

**Comment by SteveO-Cray** [comment]
> No worries. Meanwhile, I noticed that on `osx+intel compiler` we can't even resolve `MPI_IN_PLACE` and `MPI_BOTTOM`. Those two symbols can't be fixed in the same way as the rest of the symbols since they are directly passed to `wrappers_c`. I am working on the last commit to see if that is due to dynamic linker issue. You can ignore the last commit in your review for now.

That was part of what I was going to look into (sorry, pulled away again).  I wondered why MPI_IN_PLACE and MPI_BOTTOM and possibly others were not affected, I guess they are.  I will get to finishing this review at some point.

**Comment by hzhou** [comment]
> > No worries. Meanwhile, I noticed that on `osx+intel compiler` we can't even resolve `MPI_IN_PLACE` and `MPI_BOTTOM`. Those two symbols can't be fixed in the same way as the rest of the symbols since they are directly passed to `wrappers_c`. I am working on the last commit to see if that is due to dynamic linker issue. You can ignore the last commit in your review for now.
> 
> That was part of what I was going to look into (sorry, pulled away again). I wondered why MPI_IN_PLACE and MPI_BOTTOM and possibly others were not affected, I guess they are. I will get to finishing this review at some point.

F08 has two types of interfaces. For functions with choice buffers, it is directly linked to `C` wrappers in `wrappers_c/f08_cdesc.c`. For other functions, it is using Fortran wrappers in `wrappers_f/f08ts.f90`. Since `MPI_IN_PLACE` and `MPI_BOTTOM` only used in choice buffer parameters, it needs a linked C symbol that can be directly accessed in the C wrappers. For the other symbols, they are checked in the Fortran wrappers, so direct linkage can be bypassed using the C "getter" functions in this PR.

However, the `status` (`MPI_STATUS_IGNORE`, `MPI_STATUSES_IGNORE`) are used in both C wrappers (e.g. `MPI_Recv`) and Fortran Wrappers (e.g. `MPI_Wait`), so it probably needs both the C symbol linkage and C getter. This is likely broken since the beginning. I'll verify with a test case and then add an additional commit to fix it.

**Comment by hzhou** [comment]
> However, the `status` (`MPI_STATUS_IGNORE`, `MPI_STATUSES_IGNORE`) are used in both C wrappers (e.g. `MPI_Recv`) and Fortran Wrappers (e.g. `MPI_Wait`), so it probably needs both the C symbol linkage and C getter. This is likely broken since the beginning and we simply missed it due to lack of tests. I'll verify with a test case and then add an additional commit to fix it.

Actually the functions with choice buffer parameters go through both Fortran wrappers and C wrappers, so the status constants are checked in the Fortran wrappers, thus we are good. In fact, we could check `MPI_IN_PLACE` and `MPI_BOTTOM` in the Fortran wrappers as well. It maybe better since that brings consistency to how we treat these constants. If that is preferred, I can amend the current code.

**Comment by hzhou** [comment]
> In fact, we could check `MPI_IN_PLACE` and `MPI_BOTTOM` in the Fortran wrappers as well.

Alas, we can't. The C wrapper need directly access the choice buffer parameter as `CFI_cdesc_t`, thus we can't replace the constant parameter in the Fortran wrappers. The current code is the only solution.

**Comment by hzhou** [comment]
NOTE: added a commit to cleanup `MPI_F08_STATUS_IGNORE` and `MPI_F08_STATUSES_IGNORE`

test:mpich/custom/pipeline
label: osx
compiler: intel

test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
@raffenet @SteveO-Cray I don't want this PR to become too stale. Should we try to make a list of concerns (which may include "approval from Bill Long")?

**Comment by raffenet** [comment]
> @raffenet @SteveO-Cray I don't want this PR to become too stale. Should we try to make a list of concerns (which may include "approval from Bill Long")?

No other concerns from me.

**Review by SteveO-Cray — APPROVED** [review]
Apologies.  I forgot about this PR.  I have not heard any further comments/objections from BIll and I think it is long enough to wait.  I think you should proceed to merge.

**Review by raffenet — APPROVED** [review]

**Pull request merged**
